File: c++/2.cpp
#include<vector>
#include<iostream>
#include<algorithm>
#include<string>
#include<cmath>
#include<stack>
using namespace std;

class Solution {
public:
    struct ListNode{
        int val;
        ListNode *next;
        ListNode() : val(0), next(nullptr) {}
        ListNode(int x) : val(0), next(nullptr) {}
        ListNode(int x, ListNode* next) : val(0), next(nullptr) {}
    };
    //从两端向中间扩散
    int maxArea(vector<int>& height) {
        int start=0,end=height.size()-1;
        int ans=0;
        while(end > start){
            ans=max(ans,(end-start)*min(height[start],height[end]));
            if(height[start] > height[end])end--;
            else start++;
        }
        return ans;
    }
    string intToRoman(int num) {
        int values[] = {1000,900,500,400,100,90,50,40,10,9,5,4,1};
        string symbols[] = {"M","CM","D","CD","C","XC","L","XL","X","IX","V","IV","I"};
        string res="";
        int i=0;
        while(num != 0){
            while(values[i] > num){
                i++;
            }
            num -= values[i];
            res += symbols[i];
        } 
        return res;
    }
    int romanToInt(string s) {
        int commonValues[] = {1000,500,100,50,10,5,1};
        int specialValues[] = {900,400,90,40,9,4};
        string commonSymbols[] = {"M","D","C","L","X","V","I"};
        string specialSymbols[] = {"CM","CD","XC","XL","IX","IV"};
        int ans=0;
        for(int i=0;i < s.size();i++){
            int j;
            for(j=0;j < 6;j++){
                if(s.substr(i,2) == specialSymbols[j]){
                    ans+=specialValues[j];
                    i+=1;
                    break;
                }    
            }
            if(j < 6)continue;
            for(j=0;j < 7;j++){
                if(i<s.size() && s.substr(i,1) == commonSymbols[j]){
                    ans+=commonValues[j];
                    break;
                }
            }
        }
        return ans;
    }
    string longestCommonPrefix(vector<string>& strs) {
        if(strs.size() == 1)return strs.front();
        string res;
        int cnt=0;
        while(true){
            for(int i=1;i<strs.size();i++){
                if(strs[i].size() == cnt || strs[i][cnt] != strs[i-1][cnt]){
                    res=strs[i].substr(0,cnt);
                    return res;
                }
            }
            cnt++;
        }
    }
    vector<vector<int>> threeSum(vector<int>& nums) {
        std::vector<vector<int>> result;
        if (nums.size() < 3) {
            return result;
        }

        std::size_t n_size = nums.size();
        std::sort(nums.begin(), nums.end());
        for (int i = 0; i < n_size; ++i) {
            // all numbers from now on will be greater than 0, no point in continuing
            if (nums[i] > 0) break;

            // we have seen this number & combo before; skip
            if (i > 0 && nums[i] == nums[i-1]) continue;

            int left = i+1, right = n_size - 1;
            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (sum < 0) {
                    ++left;
                } else if (sum > 0) {
                    --right;
                } else {
                    result.push_back({nums[i], nums[left], nums[right]});
                    int last_left = nums[left], last_right = nums[right];
                    // we have seen this number & combo before; skip
                    while (left < right && nums[left] == last_left) ++left;
                    while (left < right && nums[right] == last_right) --right;
                }
            }

        }
        return result;
    }
    int threeSumClosest(vector<int>& nums, int target) {
        int def = 0x7fffffff;

        std::size_t n_size = nums.size();
        std::sort(nums.begin(), nums.end());
        int result = nums[0] + nums[1] + nums[2];
        for (int i = 0; i < n_size; ++i) {
            // all numbers from now on will be greater than 0, no point in continuing
            if (nums[i] > target / 3) break;

            // we have seen this number & combo before; skip
            if (i > 0 && nums[i] == nums[i-1]) continue;

            int left = i+1, right = n_size - 1;
            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                int tmp = sum - target;
                //the closest answer
                if(tmp == 0)return target;
                if(abs(tmp) < def){
                    result = sum;
                    def = abs(tmp);
                }
                if (tmp < 0) {
                    ++left;
                } else{
                    --right;
                }
            }
        }
        return result;
    }
    void helper(string digits,int depth,string tmpRes,vector<string>& res){
        if(depth == digits.size()){
            res.push_back(tmpRes);
            return;
        }
        string letters[] = {"abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        string digit = letters[(int)(digits[depth]-'2')]; 
        int len = digit.size();
        for(int i=0;i < len;i++){  
            helper(digits,depth+1,tmpRes+digit[i],res);
        }
    }
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        if(digits.empty())return res;
        helper(digits,0,"",res);
        return res;
    }

    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        if (nums.size() < 4) return {};
        vector<vector<int>> res;
        sort(nums.begin(), nums.end());
        int len = nums.size();
        for (int i = 0; i < len-3; i++) {
            if (i > 0 && nums[i] == nums[i-1])continue;
            for (int j = i+1; j < len-2; j++) {
                if (j > i+1 && nums[j] == nums[j-1])continue;
                int left = j+1, right = len-1;
                long long sum_i_j = nums[i] + nums[j];
                while (right > left) {
                    long long sum = sum_i_j - target + nums[left] + nums[right];
                    if (sum > 0) {
                        right--;
                    } else if (sum < 0) {
                        left++;
                    } else {
                        res.push_back({nums[i], nums[j], nums[left], nums[right]});
                        int last_left = nums[left], last_right = nums[right];
                        while (right > left && nums[right] == last_right) right--;
                        while (right > left && nums[left] == last_left) left++;
                    }
                }
            }
        }
        return res;
    }

    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *first=head, *second=head;
        while(n--){
            first = first->next;
        }
        // 删除头结点
        if(first == NULL) return head->next;
        while(first->next != NULL){
            first = first->next;
            second = second->next;
        }
        second->next = second->next->next;
        return head;
    }
    bool isValid(string s) {
        stack<char> sta;
        for(int i = 0; i < s.size(); i++){
            if(s[i] == '(' || s[i] == '[' || s[i] == '{'){
                sta.push(s[i]);
            }else if(!sta.size()){
                return false;
            }
            else if(s[i] == ')' && sta.top() == '(' ||
                    s[i] == ']' && sta.top() == '[' ||
                    s[i] == '}' && sta.top() == '{'){
                sta.pop();
            }else{
                return false;
            }
    	}
        if(sta.size())return false;
        return true;
    }
};

int main(){
    Solution solution=Solution();
    // cout<<solution.romanToInt("MCMXCIV");
    vector<string> strs;
    strs.push_back("flower");
    strs.push_back("flow");
    strs.push_back("flight");
    // cout << solution.longestCommonPrefix(strs);
    solution.letterCombinations("23");
    printf("run success!");

    int target = -11;
    vector<int> nums = {1,-2,-5,-4,-3,3,3,5};
    solution.fourSum(nums, target);
    return 0;
}

File: deprecated/Go/leetcode/2.go
package leetcode
import (
	"math"
	"sort"
)

func maxArea(height []int) int{
	max,start,end := 0,0,len(height)-1
	for start < end{
		width := end - start
		high := 0
		if height[start] < height[end]{
			high = height[start]
			start++
		}else{
			high = height[end]
			end--
		}

		t := width*high
		if t > max{
			max = t
		}
	}
	return max
}

func intToRoman(num int) string {
    values := []int{1000,900,500,400,100,90,50,40,10,9,5,4,1}
	symbols := []string{"M","CM","D","CD","C","XC","L","XL","X","IX","V","IV","I"}
	res,i := "",0
	for num != 0{
		for values[i] > num{
			i++
		}
		num -= values[i]
		res += symbols[i]
	} 
	return res
}

func romanToInt(s string) int {
	if s==""{
		return 0;
	}
	var roman = map[string]int{
		"I":1,
		"V":5,
		"X":10,
		"L":50,
		"C":100,
		"D":500,
		"M":1000,
	}
	num,lastint,total := 0,0,0
	for i:=0;i<len(s);i++{
		char := s[len(s)-(i+1):len(s)-i]
		num = roman[char]
		if num < lastint{
			total = total - num
		}else{
			total = total + num
		}
		lastint = num
	}
	return total
}

//最优解，双指针 + 排序
func threeSum(nums []int) [][]int {
    sort.Ints(nums)
	result,start,end,index,addNum,length := make([][]int,0),0,0,0,0,len(nums)
	for index = 1;index < length - 1;index++{
		start,end = 0,length-1
		if index > 1 && nums[index] == nums[index-1]{
			start = index -1
		}
		for start < index && end > index{
			if start > 0 && nums[start] == nums[start-1]{
				start++
				continue
			} 
			if end < length-1 && nums[end] == nums[end+1]{
				end--
				continue
			}
			addNum = nums[start] + nums[end] + nums[index]
			if addNum == 0{
				result = append(result,[]int{nums[start],nums[index],nums[end]})
				start++
				end--
			}else if addNum > 0{
				end--
			}else{
				start++
			}
		}
	}
	return result
}

//解法二
func threeSum1(nums []int) [][]int{
	res := [][]int{}
	counter := map[int]int{}
	for _,value := range nums{
		counter[value]++
	}

	uniqNums := []int{}
	for key := range counter{
		uniqNums = append(uniqNums,key)
	}
	sort.Ints(uniqNums)

	for i := 0;i < len(uniqNums);i++{
		if (uniqNums[i]*3 == 0) && counter[uniqNums[i]] >= 3{
			res = append(res,[]int{uniqNums[i],uniqNums[i],uniqNums[i]})
		}
		for j := i+1;j < len(uniqNums);j++{
			if (uniqNums[i]*2+uniqNums[j] == 0) && counter[uniqNums[i]] > 1{
				res = append(res,[]int{uniqNums[i],uniqNums[i],uniqNums[j]})
			}
			if (uniqNums[j]*2+uniqNums[i] == 0) && counter[uniqNums[j]] > 1{
				res = append(res,[]int{uniqNums[i],uniqNums[j],uniqNums[j]})
			}
			c := 0-uniqNums[i] - uniqNums[j]
			if c > uniqNums[j] && counter[c] > 0{
				res =  append(res,[]int{uniqNums[i],uniqNums[j],c})
			}
		}
	}
	return res
}

//解法一O(n^2)
func threeSumClosest(nums []int,target int)int{
	n,res,diff := len(nums),0,math.MaxInt32
	if n > 2{
		sort.Ints(nums)
		for i := 0;i < n-2;i++{
			for j,k := i+1,n-1;j<k; {
				sum := nums[i] + nums[j] + nums[k]
				if abs(sum-target) < diff{
					res,diff = sum,abs(sum-target)
				}
				if sum == target{
					return res
				}else if sum > target{
					k--
				}else{
					j++
				}
			}
		}
	}
	return res
}

//解法二 暴力解法O(n^3)
func threeSumClosest1(nums []int,target int)int{
	res,difference := 0,math.MaxInt16
	for i:=0;i<len(nums);i++{
		for j:=i+1;j<len(nums);j++{
			for k:=j+1;k<len(nums);k++{
				if abs(nums[i]+nums[j]+nums[k]-target) < difference{
					difference = abs(nums[i]+nums[j]+nums[k]-target)
					res = nums[i] + nums[j] + nums[k]
				}
			}
		}
	}
	return res
}
func abs(a int)int{
	if a>0{
		return a
	}
	return -a
}


var (
	letterMap = []string{
		" ",
		"",
		"abc",
		"def",
		"ghi",
		"jkl",
		"mno",
		"pqrs",
		"tuv",
		"wxyz",
	}
	res = []string{}
	final = 0
)

//解法一DFS
func letterCombinations(digits string) []string{
	if digits == ""{
		return []string{}
	}
	res = []string{}
	findCombination(&digits,0,"")
	return res
}

func findCombination(digits *string,index int,s string){
	if index == len(*digits){
		res = append(res,s)
		return 
	}
	num := (*digits)[index]
	letter := letterMap[num-'0']
	for i:=0;i<len(letter);i++{
		findCombination(digits,index+1,s+string(letter[i]))
	}
	return 
}

//解法二 非递归
func letterCombinations_(digits string) []string{
	if digits == ""{
		return []string{}
	}
	index := digits[0]-'0'
	letter := letterMap[index]
	tmp := []string{}
	for i:=0;i<len(letter);i++{
		if len(res) == 0{
			res = append(res,"")
		}
		for j:=0;j<len(res);j++{
			tmp = append(tmp,res[j]+string(letter[i]))
		}
	}
	res = tmp
	final++
	letterCombinations_(digits[1:])
	final--
	if final == 0{
		tmp = res
		res = []string{}
	}
	return tmp
}

//解法三 回溯
var result []string
var dict = map[string][]string{
	"2" : []string{"a","b","c"},
	"3" : []string{"d","e","f"},
	"4" : []string{"g","h","i"},
	"5" : []string{"j","k","l"},
	"6" : []string{"m","n","o"},
	"7" : []string{"p","q","r","s"},
	"8" : []string{"t","u","v"},
	"9" : []string{"w","x","y","z"},
}

func letterCombinationsBT(digits string)[]string{
	result = []string{}
	if digits == ""{
		return result
	}
	letterFunc("",digits)
	return result	
}

func letterFunc(res string,digits string){
	if digits == ""{
		result = append(result,res)
		return 
	}
	 
	k := digits[0:1]
	digits = digits[1:]
	for i:=0;i<len(dict[k]);i++{
		res += dict[k][i]
		letterFunc(res,digits)
		res = res[0:len(res)-1]
	}
}